namer/namer.cc

```diff
@@ -15,16 +18,1336 @@
+}
+const FoundStaticField &FoundNameRef::staticField(const FoundNames &foundNames) const {
+    ENFORCE(kind() == NameKind::StaticField);
+    ENFORCE(foundNames._staticFields.size() > idx());
+    return foundNames._staticFields[idx()];
+}
+
+FoundTypeMember &FoundNameRef::typeMember(FoundNames &foundNames) {
+    ENFORCE(kind() == NameKind::TypeMember);
+    ENFORCE(foundNames._typeMembers.size() > idx());
+    return foundNames._typeMembers[idx()];
+}
+const FoundTypeMember &FoundNameRef::typeMember(const FoundNames &foundNames) const {
+    ENFORCE(kind() == NameKind::TypeMember);
+    ENFORCE(foundNames._typeMembers.size() > idx());
+    return foundNames._typeMembers[idx()];
+}
+
+core::SymbolRef FoundNameRef::symbol(const FoundNames &foundNames) const {
+    ENFORCE(kind() == NameKind::Symbol);
+    ENFORCE(foundNames._symbols.size() > idx());
+    return foundNames._symbols[idx()];
+}
+
+struct SymbolFinderResult {
+    ast::ParsedFile tree;
+    unique_ptr<FoundNames> names;
+};
+
+core::SymbolRef methodOwner(core::Context ctx, const ast::MethodDef::Flags &flags) {
+    core::SymbolRef owner = ctx.owner.data(ctx)->enclosingClass(ctx);
+    if (owner == core::Symbols::root()) {
+        // Root methods end up going on object
+        owner = core::Symbols::Object();
+    }
+
+    if (flags.isSelfMethod) {
+        if (owner.data(ctx)->isClassOrModule()) {
+            owner = owner.data(ctx)->lookupSingletonClass(ctx);
+        }
+    }
+    ENFORCE(owner.exists());
+    ENFORCE(owner.data(ctx)->isClassOrModule());
+    return owner;
+}
+
+// Returns the SymbolRef corresponding to the class `self.class`, unless the
+// context is a class, in which case return it.
+core::SymbolRef contextClass(const core::GlobalState &gs, core::SymbolRef ofWhat) {
+    core::SymbolRef owner = ofWhat;
+    while (true) {
+        ENFORCE(owner.exists(), "non-existing owner in contextClass");
+        const auto &data = owner.data(gs);
+
+        if (data->isClassOrModule()) {
+            break;
+        }
+        if (data->name == core::Names::staticInit()) {
+            owner = data->owner.data(gs)->attachedClass(gs);
+        } else {
+            owner = data->owner;
+        }
+    }
+    return owner;
+}
+
+/**
+ * Used with TreeMap to locate all of the class, method, static field, and type member symbols defined in the tree.
+ * Does not mutate GlobalState, which allows us to parallelize this process.
+ * Does not report any errors, which lets us cache its output.
+ * Produces a vector of symbols to insert, and a vector of modifiers to those symbols.
+ */
+class SymbolFinder {
+    unique_ptr<FoundNames> foundNames = make_unique<FoundNames>();
+    // The tree doesn't have symbols yet, so `ctx.owner`, which is a SymbolRef, is meaningless.
+    // Instead, we track the owner manually as an index into `foundNames`; the item at that index
+    // will define the `owner` symbol.
+    // The u4 is actually index + 1 so we reserve 0 for the root owner.
```

Comment:
Is this comment about `index + 1` still valid? I couldn't track down what this is referring to. It looks like `getOwner()` returns `root()` when the stack is empty, rather than always assuming the stack is non-empty and returning `.back()` unconditionally.

Reply:
This comment is out-of-date.